Add conmon man page (#210)
containers/conmon: docs/conmon.8.md

conmon 8 "User Commands"
==================================================

# NAME

conmon - container monitor utility

# SYNOPSIS

conmon [options]

# DESCRIPTION

conmon is a command line program for monitorying and managing the lifecycle of

> linux containers follow the Open Container Initiative (OCI) format.

Reviewer: ```suggestion
Linux containers that follow the Open Container Initiative (OCI) format.
```

> # APPLICATION OPTIONS

Reviewer: Need periods at the end fo lines.  Many of the descriptions are basically a rehash of the option name and need more verbiage.

Author: These all come from the output of conmon --help.  So we really need to fix it in both places.

**--api-version**
Conmon API version to use

**-b**, **--bundle**
Bundle path

**-c**, **--cid**
Container ID

**--exec-attach**
Attach to an exec session

**-e**, **--exec**
Exec a command in a running container

**--exec-process-spec**
Path to the process spec for exec

**--exit-command**
Path to the program to execute when the container terminates its execution

**--exit-command-arg**

> Additional arg to pass to the exit command.  Can be specified multiple times

Reviewer: ```suggestion
Additional argument to pass to the exit command.  Can be specified multiple times
```

**--exit-delay**
Delay before invoking the exit command (in seconds)

**--exit-dir**
Path to the directory where exit files are written

**-h**, **--help**
Show help options

**-i**, **--stdin**
stdin

**-l**, **--log-path**
Log file path

**--leave-stdin-open**
Leave stdin open when attached client disconnects

**--log-level**
Print debug logs based on log level

**--log-size-max**
Maximum size of log file

**--log-tag**
Additional tag to use for logging

**-n**, **--name**
Container name

**--no-new-keyring**
Do not create a new session keyring for the container

**--no-pivot**
Do not use pivot_root

**--no-sync-log**
Do not manually call sync on logs after container shutdown

**-0**, **--persist-dir**
Persistent directory for a container that can be used for storing container data

**--pidfile**
PID file (DEPRECATED)

**-p**, **--container-pidfile**
Container PID file

**-P**, **--conmon-pidfile**
Conmon daemon PID file

**-r**, **--runtime**
Runtime path

**--replace-listen-pid**
Replace listen pid if set for oci-runtime pid

**--restore**
Restore a container from a checkpoint

**--restore-arg**

> Additional arg to pass to the restore command. Can be specified multiple times. (DEPRECATED)

Reviewer: ```suggestion
Additional argument to pass to the restore command. Can be specified multiple times. (DEPRECATED)
```

**--runtime-arg**

> Additional arg to pass to the runtime. Can be specified multiple times

Reviewer: ```suggestion
Additional argument to pass to the runtime. Can be specified multiple times
```

**--runtime-opt**

> Additional opts to pass to the restore or exec command. Can be specified multiple times

Reviewer: ```suggestion
Additional options to pass to the restore or exec command. Can be specified multiple times
```

**-s**, **--systemd-cgroup**

> Enable systemd cgroup manager

Reviewer: which allows?

Author: @giuseppe @vrothberg @haircommander 
Does this just tell conmon to communicate with systemd as opposed to use cgroupfs directly?

**--socket-dir-path**
Location of container attach sockets

**--sync**
Allowing caller to keep the main conmon process as its child by only forking once

**--syslog**
Log to syslog (use with cgroupfs cgroup manager)

**-t**, **--terminal**

> Terminal

Reviewer: can we expound a bit more on this?

Author: Sure I have no idea what this means.  @haircommander Below there is another field on hold open standard IO?

Does this mean the same thing as podman run -t?

```
       Allocate a pseudo-TTY. The default is false.

       When set to true, Podman will allocate a pseudo-tty and attach to the standard input of the container. This can be used, for example, to run a throwaway interactive shell. The default is false.

       NOTE: The -t option is incompatible with a redirection of the Podman client standard input.
```

Reviewer 2: yeah conmon has a `-i` that corresponds to podman's stdin.

`-t` should be called when the manager wants a terminal connection. Conmon will open a pseudoterminal console, and pass it to the runtime with `--console-socket`

conmon 8 "User Commands"
==================================================

# NAME

conmon - container monitor utility

# SYNOPSIS

conmon [options]

# DESCRIPTION

> conmon is a command line program for monitorying and managing the lifecycle of

Reviewer: ```suggestion
conmon is a command-line program for monitoring and managing the lifecycle of
```

Linux containers that follow the Open Container Initiative (OCI) format.

# APPLICATION OPTIONS

**--api-version**
Conmon API version to use.

**-b**, **--bundle**
Location of the OCI Bundle path.

**-c**, **--cid**
Identification of Container.

**--exec-attach**
Attach to an exec session.

**-e**, **--exec**
Exec a command into a running container.

**--exec-process-spec**
Path to the process spec for execution.

**--exit-command**
Path to the program to execute when the container terminates its execution.

**--exit-command-arg**
Additional arguments to pass to the exit command.  Can be specified multiple time.

**--exit-delay**
Delay before invoking the exit command (in seconds).

**--exit-dir**
Path to the directory where exit files are written.

**-h**, **--help**
Show help options.

**-i**, **--stdin**
Open up a pipe to pass stdin to the container.

> This flag tells conmon to setup the pipe regardless of whether there is a terminal connection.

Reviewer: ```suggestion
This option tells conmon to setup the pipe regardless of whether there is a terminal connection.
```

**-l**, **--log-path**
Path to store all stdout and stderr messages from the container.

**--leave-stdin-open**
Leave stdin open when the attached client disconnects.

**--log-level**
Print debug logs based on the log level.

**--log-size-max**
Maximum size of the log file.

**--log-tag**

> Additional tag to use for logging.

Reviewer: IDK, can thist be specified multiple times?  If so we should note that.

**-n**, **--name**
Container name.

**--no-new-keyring**
Do not create a new session keyring for the container.

**--no-pivot**
Do not use pivot_root.

**--no-sync-log**
Do not manually call sync on logs after container shutdown.

**-0**, **--persist-dir**
Persistent directory for a container that can be used for storing container data.

**-p**, **--container-pidfile**

> PID file for the initial pid inside of container.

Reviewer: nit
```suggestion
PID file for the initial pid inside of the container.
```

**-P**, **--conmon-pidfile**
PID file for the conmon process.

**-r**, **--runtime**
Path to store runtime data for the container.

**--replace-listen-pid**
Replace listen pid if set for oci-runtime pid.

**--restore**
Restore a container from a checkpoint.

**--runtime-arg**
Additional arguments to pass to the runtime. Can be specified multiple times.

**--runtime-opt**
Additional options to pass to the restore or exec command. Can be specified multiple times.

**-s**, **--systemd-cgroup**
Enable systemd cgroup manager, rather then use the cgroupfs directly.

**--socket-dir-path**
Location of container attach sockets.

**--sync**
Keep the main conmon process as its child by only forking once.

**--syslog**
Log to syslog (use with cgroupfs cgroup manager).

**-t**, **--terminal**
Allocate a pseudo-TTY. The default is false.

When set to true, conmon will allocate a pseudo-tty and attach  to  the
standard  input of the container. This can be used, for example, to run
a throwaway interactive shell. The default is false.

> **-5** should be called when the manager wants a terminal connection. Conmon will open a pseudoterminal console, and pass it to the OCI runtime with **--console-socket**.

Reviewer: is this option really `5`?  If so, it should be the first option in this list of options?

conmon 8 "User Commands"
==================================================

# NAME

conmon - container monitor utility

# SYNOPSIS

conmon [options]

# DESCRIPTION

conmon is a command line program for monitorying and managing the lifecycle of
Linux containers that follow the Open Container Initiative (OCI) format.

# APPLICATION OPTIONS

**--api-version**
Conmon API version to use.

**-b**, **--bundle**
Location of the OCI Bundle path.

**-c**, **--cid**
Identification of Container.

**--exec-attach**
Attach to an exec session.

**-e**, **--exec**
Exec a command into a running container.

**--exec-process-spec**
Path to the process spec for execution.

**--exit-command**
Path to the program to execute when the container terminates its execution.

**--exit-command-arg**
Additional arg to pass to the exit command.  Can be specified multiple time.

**--exit-delay**
Delay before invoking the exit command (in seconds).

**--exit-dir**
Path to the directory where exit files are written.

**-h**, **--help**
Show help options.

**-i**, **--stdin**

> Create !terminal stdin pipe.

Reviewer: ```suggestion
Create a terminal stdin pipe.
```

Author: That came from the code, Not sure what it means, so I want @haircommander To comment.

Reviewer 2: it should actually be:
Open up a pipe to pass stdin to the container, regardless of whether there is a terminal connection.

Author: Fixed.

> Create a "fake" main fd so that we can use the same epoll code in both cases.

Reviewer: ```suggestion
Create a "fake" main file descriptor so that we can use the same epoll code in both cases.
```
I'm not sure why a user would care about the epoll code or even what it is potentially.  Does this sentence even belong here?

The workerfd_*s will be closed after we dup over everything. We use pipes here because
open(/dev/std{out,err}) will fail if we used anything else (and it wouldn't be a good idea to

> create a new pty pair in the host).

Reviewer: I'd drop the conversational tone and the we/you use in this section.

Reviewer: Re-reading, this all sounds like a good code comment, not a description of the functionality from a user perspective.

Author: Yes Want @haircommander to rewrite this section for Human consumption.

Reviewer 2: I think we can just drop this whole section

Author: Removed

**-l**, **--log-path**
Path to store all stdout and stderr messages from the container.

**--leave-stdin-open**

> Leave stdin open when attached client disconnects.

Reviewer: ```suggestion
Leave stdin open when the attached client disconnects.
```

**--log-level**

> Print debug logs based on log level.

Reviewer: ```suggestion
Print debug logs based on the log level.
```

**--log-size-max**

> Maximum size of log file.

Reviewer: ```suggestion
Maximum size of the log file.
```

**-P**, **--conmon-pidfile**
PID file for the conmon process.

**-r**, **--runtime**
Path to store runtime data for the container.

**--replace-listen-pid**
Replace listen pid if set for oci-runtime pid.

**--restore**
Restore a container from a checkpoint.

**--runtime-arg**
Additional argument to pass to the runtime. Can be specified multiple times.

**--runtime-opt**
Additional opts to pass to the restore or exec command. Can be specified multiple times.

**-s**, **--systemd-cgroup**
Enable systemd cgroup manager, rather then use the cgroupfs directly.

**--socket-dir-path**
Location of container attach sockets.

**--sync**
Keep the main conmon process as its child by only forking once.

**--syslog**
Log to syslog (use with cgroupfs cgroup manager).

**-t**, **--terminal**
Allocate a pseudo-TTY. The default is false.

When set to true, conmon will allocate a pseudo-tty and attach  to  the
standard  input of the container. This can be used, for example, to run
a throwaway interactive shell. The default is false.

**-T**, **--timeout**
Kill container after specified timeout in seconds.

**-u**, **--cuuid**
Specify the Container UUID to use.

**--version**

> Print the version and exit.

Reviewer: I'd change "opt" and "arg" to "option" and "argument" or their plural throughout.

Author: Done.
